src/Microsoft.AspNet.OData.Shared/Query/ODataPathQueryBuilder.cs

```diff
@@ -0,0 +1,210 @@
+                Expression.Lambda(propertyExpression, param);
+            return ExpressionHelpers.Select(result, selectBody, queryable.ElementType);
+        }
+
+        /// <summary>
+        /// Transforms <paramref name="queryable"/> by selecting
+        /// the collection-valued property called <paramref name="propertyName"/>.
+        /// </summary>
+        /// <param name="queryable">The <see cref="IQueryable"/> to apply the property selection to</param>
+        /// <param name="propertyName">The name of the collection-valued property to select</param>
+        /// <returns>The transformed <see cref="IQueryable"/></returns>
+        private static IQueryable GetCollectionPropertyQuery(IQueryable queryable, string propertyName)
+        {
+            ParameterExpression param = Expression.Parameter(queryable.ElementType);
+            MemberExpression propertyExpression = Expression.Property(param, propertyName);
+
+            // for collection properties we use SelectMany instead of Select
+            // because Select would return an IQueryable<IEnumerable<TResult>>,
+            // but SelectMany flattens it to IQueryable<TResult>
+
+            // we don't check whether the property is null before advancing
+            // because that seems to cause exceptions in EF
+
+            // We expect the collection navigation property to implements IEnumerable<T>
+            // Here we extract the element type T
+            Type collectionPropertyElementType = TypeHelper.GetImplementedIEnumerableType(propertyExpression.Type) ?? propertyExpression.Type;
+
+            // The lambda passed to queryable.SelectMany() has a signature Func<TSource, IEnumerable<TResult>>
+            // TSource corresponds to our queryable.ElementType and TResult is the element type of the nav property
+            Type delegateType = typeof(Func<,>).MakeGenericType(
+                queryable.ElementType,
+                typeof(IEnumerable<>).MakeGenericType(collectionPropertyElementType));
+            LambdaExpression selectBody =
+                Expression.Lambda(delegateType, propertyExpression, param);
+
+            // return queryable.SelectMany(entity => entity.CollectionNavProp)
+            return ExpressionHelpers
+                .SelectMany(queryable, selectBody, queryable.ElementType, collectionPropertyElementType);
+        }
+    }
+
+    /// <summary>
+    /// Result returned by <see cref="ODataPathQueryBuilder"/> after
+    /// applying transformations based on path.
+    /// </summary>
+    internal class ODataPathQueryResult
+    {
+        /// <summary>
+        /// The result of the query transformations applied to the original <see cref="IQueryable"/>
+        /// by the <see cref="ODataPathQueryBuilder"/>
+        /// </summary>
+        public IQueryable Result { get; set; }
+        /// <summary>
+        /// Whether the path has a $count segment
+        /// </summary>
+        public bool HasCountSegment { get; set; }
```

Comment:
How is this property used in computing the count value?

Reply:
This flag is set to true if the path contains a `$count` segment. This informs as that the result of the query should be treated as a collection.